core/src/main/java/com/graphhopper/routing/weighting/custom/CustomWeightingHelper.java

```diff
@@ -0,0 +1,65 @@
+/*
+ *  Licensed to GraphHopper GmbH under one or more contributor
+ *  license agreements. See the NOTICE file distributed with this work for
+ *  additional information regarding copyright ownership.
+ *
+ *  GraphHopper GmbH licenses this file to you under the Apache License,
+ *  Version 2.0 (the "License"); you may not use this file except in
+ *  compliance with the License. You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+package com.graphhopper.routing.weighting.custom;
+
+import com.graphhopper.json.geo.JsonFeature;
+import com.graphhopper.routing.ev.DecimalEncodedValue;
+import com.graphhopper.routing.ev.EncodedValueLookup;
+import com.graphhopper.util.EdgeIteratorState;
+import com.graphhopper.util.FetchMode;
+import com.graphhopper.util.GHUtility;
+import com.graphhopper.util.shapes.BBox;
+import com.graphhopper.util.shapes.Polygon;
+
+import java.util.Map;
+
+/**
+ * This class makes it easy to control priority and speed for the CustomWeighting.
+ */
+public class CustomWeightingHelper {
```

Comment:
Ah ok the default constructor + init method is probably required to create an instance via reflection in ExpressionBuilder? But to me this sounds like just another reason to have some kind of `SpeedAndPriorityProvider` interface in `CustomWeighting` and encapsulate the ugly part of creating such a thing from a custom model. I mean `CustomWeighting` should not be coupled to this ugly part? One goal would be to be able to test `CustomModel`->janino->`SpeedAndPriortyProvider` independently from `SpeedAndPriorityProvider`->`CustomWeighting`->`calcWeight`.

Reply:
I can try to make it private or protected and see if Janino can work with it. It should not be used from outside - only for internal purposes.

> The static in method seems very unrelated and can probably go to GHUtilities?

IMO it could make sense to have all methods that we allow for scripting there directly instead of GHUtilities?

> Ah ok the default constructor + init method is probably required to create an instance via reflection in ExpressionBuilder?

Yes, I was just too lazy too look how to call the correct constructor and used the explicit init method.

> One goal would be to be able to test 

Not sure if this is necessary. What purpose would this have?

Reply:
> IMO it could make sense to have all methods that we allow for scripting there directly instead of GHUtilities?

Yes if its specific to custom model scripting better keep it closer to that code, but the method that decides whether an edge lies within a polygon seems general and useful enough to go to GHUtilities?

> Not sure if this is necessary. What purpose would this have?

To me there just seem to be two 'stories' here and it makes the code easier to understand, test and change when we keep them separated. One story is about interpreting our custom model language and extracting a mapping between edges and speeds/priorities. This is already quite a complex task which involves janino and giving a meaning to the keys and values we define in the custom model. The other story is about taking the edge->speed/priority mapping and calculating edge weights from this (applying the formula we use in CustomWeighting). To me this seems to be independent from creating this mapping and for example we should be able to use CustomWeighting with some mapping that does not use yaml, janino and our custom model language at all. This also becomes clear when we think about testing : Right now its hard to test the first story in isolation (and I guess this is why we do not have such tests yet). We cannot really test the custom model language, janino and all this without creating a CustomWeighting. We also cannot create a CustomWeighting without using ExpressionBuilder or CustomWeightingHelper which are both related to janino. At the same time this coupling seems easily avoidable, because the edge->speed/priority mapping provides a very small boundary between the CustomWeighting and the custom model language interpreting code.

Have a look at this test in `CustomWeightingTest`:

```java
       EdgeIteratorState primary = graph.edge(0, 1).setDistance(10).
                set(roadClassEnc, PRIMARY).set(avSpeedEnc, 80).set(accessEnc, true, true);
        EdgeIteratorState secondary = graph.edge(1, 2).setDistance(10).
                set(roadClassEnc, SECONDARY).set(avSpeedEnc, 70).set(accessEnc, true, true);

        CustomModel vehicleModel = new CustomModel();
        vehicleModel.getPriority().put("road_class != PRIMARY", 0.5);

        assertEquals(1.15, createWeighting(vehicleModel).calcEdgeWeight(primary, false), 0.01);
        assertEquals(1.73, createWeighting(vehicleModel).calcEdgeWeight(secondary, false), 0.01);
```
Here we create some edges and a custom model and what we really want to do is testing if our custom model rules yield the right speed/priority. But we cannot test this directly so we create a CustomWeighting and have to check the outcome of our custom model indirectly (we have to know the formula in CustomWeighting). So we need to compare the weight we get to some magic numbers like 1.15 and 1.73.

I will try to refactor this a bit so you can see what I mean exactly.

Reply:
See this draft: https://github.com/graphhopper/graphhopper/pull/2211